File: src/shape.cpp
#include <cell_world/shape.h>
#include <cell_world/geometry.h>

namespace cell_world {
    //SHAPE

    Shape::Shape() = default;

    Shape::Shape(int sides):
            sides(sides){

    }

    //POLYGON

    Polygon::Polygon() = default;

    Polygon::Polygon(const Location &center, unsigned int sides, float radius, float rotation) :
            center(center),
            radius(radius){
        float theta = to_radians(rotation);
        float inc = to_radians(360.0 / float(sides));
        for (unsigned int s = 0; s < sides; s++) {
            auto c = center;
            vertices.emplace_back(c.move(theta, radius));
            theta += inc;
        }
    }

    Polygon::Polygon(const Location &center,const Shape &shape, const Transformation &transformation) :
            Polygon(center, shape.sides, transformation.size / 2, transformation.rotation)
    {

    }


    bool Polygon::contains(const Location &location) const {
        auto dist = center.dist(location);
        if (dist>radius) return false;
        auto inner_radius = center.dist(vertices[0],vertices[1]);
        if (dist<inner_radius) return true;
        for (unsigned int i=1; i<vertices.size(); i++)
            if (segments_intersect(center,location,vertices[i],vertices[i-1]))
                return false;
        if (segments_intersect(center,location,vertices[vertices.size()-1],vertices[0]))
            return false;
        return true;
    }


    Polygon Polygon::move(const Location &dst) const {
        Polygon p;
        p = *this;
        auto dif = dst - center;
        return p += dif;
    }

    Polygon Polygon::move(float theta, float dist) const {
        return move(center + Location().move(theta,dist));
    }

    Polygon Polygon::operator+=(const Location &dif) {
        center = center + dif;
        radius = radius;
        for (auto &v: vertices){
            v += dif;
        }
        return *this;
    }

    Polygon &Polygon::operator=(const Polygon &poly) {
        center = poly.center;
        radius = poly.radius;
        vertices = poly.vertices;
        return *this;
    }

    Polygon::Polygon(const Polygon &p) {
        *this = p;
    }

    bool Polygon::is_between(const Location &src, const Location &dst) const {
        float theta = src.atan(dst);
        float dist = src.dist(dst);
        return is_between(src, theta, dist);
    }

    bool Polygon::is_between(const Location &src, float theta, float dist) const {
        float dist_center = src.dist(center);
        if (dist < dist_center - radius ) return false; // destination closer than the polygon

        float polygon_shade_angle = atan2(dist_center,  radius);
        float theta_center = src.atan(center);
        auto diff_theta_center = angle_difference(theta, theta_center);
        if (diff_theta_center > polygon_shade_angle) return false; // outside the circular shade

        auto direction_center = direction(theta, theta_center);
        for (auto &v: vertices) {
            float vertex_distance = src.dist(v);
            if (vertex_distance < dist) {
                float theta_vertex = src.atan(v);
                auto direction_vertex = direction(theta, theta_vertex);
                if (direction_center == -direction_vertex) {
                    auto diff_theta_vertex = angle_difference(theta,theta_vertex);
                    if (diff_theta_center + diff_theta_vertex < M_PI)
                        return true;
                }
            }
        }
        return false;
    }

    bool Polygon::contains(const Polygon &polygon) const {
        for(auto &v:polygon.vertices)
            if (!contains(v)) return false;
        return true;
    }

    bool Polygon::overlaps(const Polygon &polygon) const {
        for(auto &v:polygon.vertices)
            if (contains(v)) return true;
        return false;
    }

    //POLYGON LIST

    Polygon_list::Polygon_list(const Location_list &centers, const Shape &shape, const Transformation &transformation) {
        for (auto &center:centers ) emplace_back(center, shape, transformation);
    }

    bool Polygon_list::contains(const Location &location) const {
        for (auto &p : *this) if (p.contains(location)) return true;
        return false;
    }

    //SPACE

    Space::Space() = default;

    Space::Space(const Location &center, const Shape &shape, const Transformation &transformation) :
            center(center),
            shape(shape),
            transformation(transformation){

    }

    Location Space::transform(const Location &location, const Space &src_space) const {
        auto source_dist_center = src_space.center.dist(location);
        auto source_theta = src_space.center.atan(location);
        auto size_ratio = transformation.size / src_space.transformation.size;
        auto rotation = to_radians(transformation.rotation - src_space.transformation.rotation);
        auto new_location = center;
        return new_location.move(rotation+source_theta, source_dist_center * size_ratio);
    }

    Location Space::scale(const Location &location, const Scale &scale) {
        return (location - center) * scale + center;
    }

}

File: python/cellworld1/src/shape.py
from json_cpp import JsonObject, JsonList
from .location import Location, Location_list, segments_intersect, Scale
from .util import *
import math


class Shape(JsonObject):
    def __init__(self, sides=6):
        self.sides = int(sides)


class Transformation(JsonObject):
    def __init__(self, size=0.0, rotation=0.0):
        self.size = float(size)
        self.rotation = float(rotation)


class Transformation_list(JsonList):
    def __init__(self, iterable=None, n=None, size=0.0, rotation=0.0):
        JsonList.__init__(self, iterable, list_type=Transformation)
        if n:
            for a in range(n):
                self.append(Transformation(float(size), float(rotation) + float(a) * (360.0 / float(n))))


class Space(JsonObject):
    def __init__(self,
                 center: Location = None,
                 shape: Shape = None,
                 transformation: Transformation = None):
        self.center = center if center else Location()
        self.shape = shape if shape else Shape()
        self.transformation = transformation if transformation else Transformation()

    @staticmethod
    def transform_to(location: Location, src_space, dst_space):
        check_type(src_space, Space, "incorrect type for src_space")
        check_type(dst_space, Space, "incorrect type for dst_space")
        size_ratio = dst_space.transformation.size / src_space.transformation.size
        rotation = math.radians(dst_space.transformation.rotation - src_space.transformation.rotation)
        dist = src_space.center.dist(location)
        theta = src_space.center.atan(location)
        new_location = Location(dst_space.center.x, dst_space.center.y)
        return new_location.move(theta + rotation, dist * size_ratio)

    def scale(self, location:Location, scale: Scale):
        diff = (location - self.center)
        scaled = diff * scale
        new_location = scaled + self.center
        return new_location

class Polygon:

    def __init__(self,
                 center: Location,
                 sides: int = 0,
                 radius: float = 1.0,
                 rotation: float = 0.0,
                 vertices: Location_list = None):
        if vertices is None:
            vertices = Location_list()
        self.center = center
        self.vertices = vertices
        self.radius = float(radius)
        if sides > 0:
            if len(self.vertices) != 0:
                raise "cannot use sides and vertices together"
            rotation = float(rotation)
            if sides == 0:
                raise "incorrect parameters"
            theta = math.radians(rotation)
            inc = 2.0 * math.pi / sides
            for s in range(sides):
                c = center.copy()
                self.vertices.append(c.move(theta, radius))
                theta += inc
        else:
            if len(self.vertices) == 0:
                raise "must specify either sides or vertices"

    def move(self,
             location: Location = None,
             theta: float = None,
             dist: float = None):
        dif = Location(0, 0)
        if location is None:
            if theta is None or dist is None:
                raise "incorrect parameters"
            dif.move(theta, dist)
        else:
            dif = location - self.center
        self.center = location
        for v in self.vertices:
            v = v + dif

    def contains(self, location: Location):
        dist = self.center.dist(location)
        if dist > self.radius:
            return False
        inner_radius = self.center.dist(segment=(self.vertices[0], self.vertices[1]))
        if dist < inner_radius:
            return True
        for i in range(1, len(self.vertices)):
            if segments_intersect((self.center, location), (self.vertices[i], self.vertices[i-1])):
                return False
        if segments_intersect((self.center, location), (self.vertices[-1], self.vertices[0])):
            return False
        return True

    def is_between(self,
                   src: Location = None,
                   dst: Location = None,
                   theta: float = None,
                   dist: float = None):
        if dst is not None:
            theta = src.atan(dst)
            dist = src.dist(dst)
        else:
            if theta is None or dist is None:
                raise "either dst or theta and dist should be used"
        dist_center = src.dist(self.center)
        theta_center = src.atan(self.center)
        diff_theta_center, direction_center = angle_difference(theta, theta_center)
        if dist < dist_center - self.radius:
            return False
        for v in self.vertices:
            vertex_distance = src.dist(v)
            if vertex_distance < dist:
                theta_vertex = src.atan(v)
                diff_theta_vertex, direction_vertex = angle_difference(theta, theta_vertex);
                if direction_center == -direction_vertex:
                    if diff_theta_center + diff_theta_vertex < math.pi:
                        return True
        return False


class Polygon_list(JsonList):

    def __init__(self, iterable=None):
        JsonList.__init__(self, iterable, list_type=Polygon)

    @staticmethod
    def get_polygons(centers: Location_list, sides: int, radius: float, rotation: float):
        polygon_list = Polygon_list()
        for center in centers:
            polygon_list.append(Polygon(center,sides=sides, rotation=rotation, radius=radius))
        return polygon_list
